### PR TITLE
Adding Detekt in the project

### PR DESCRIPTION
**Pull request by theolm**
# Description:

This pull request aims to enhance the quality of our source code by replacing the ktlint code formatter with Detekt. Detekt is a powerful static analysis tool for Kotlin, providing comprehensive analyses to improve code readability and maintainability. Here are some of the benefits of using this tool:

1. **Comprehensive Static Analysis:**
   - Detekt goes beyond basic formatting, offering more extensive static analyses to identify code patterns, anti-patterns, and other issues that can impact code quality.

2. **Flexible Configuration:**
   - Detekt allows flexible configuration, adapting to the specific needs of our project. I already added all the ktlint rules and Compose rules so we don't need to rely on other linters.

3. **Integration with Android Studio:**
   - Detekt provides direct integration with Android Studio via a plugin, enhancing the development experience for our Kotlin-based Android projects. When setting up the plugin, please don't forget to use our baseline and config, otherwise it will not work properly. 

4. **Baseline for Existing Errors:**
   - A baseline has been added to prevent detection of existing errors, allowing for a smoother transition. This ensures that the current codebase's issues are not flagged immediately.

## Other changes
**Workflow Updates:**
   - Existing workflows have been updated to incorporate the changes related to Detekt integration, ensuring a seamless integration into our development processes.

## Future work
In a subsequent PR, the baseline will be removed, and existing errors will be addressed. This phased approach allows us to adopt Detekt gradually without overwhelming the team with immediate corrections.

## How to Test:

1. Ensure the environment is correctly configured after merging this PR.
2. Run `./gradlew detekt` to ensure Detekt's static analyses are executed successfully.

## More info:

Here is a report for all the issues found by the tool.

[report.zip](https://github.com/mihonapp/mihon/files/14030252/report.zip)

Please let me know what you guys think.

### CODE REVIEW

**Comment by AntsyLich** [comment]
Is detekt being applied to all modules?

**Comment by theolm** [comment]
The plugin is applied only to the app module, but the configuration is set to check files across the entire project (all the modules). I wrote a pre-compiled plugin with the detekt configuration. `buildSrc/src/main/kotlin/detekt.gradle.kts`

```kt
private val projectSource = file("$rootDir")
private val configFile = files("$rootDir/config/detekt/detekt.yml")
private val baselineFile = file("$rootDir/config/detekt/baseline.xml")
private val kotlinFiles = "**/*.kt"
private val resourceFiles = "**/resources/**"
private val buildFiles = "**/build/**"
private val generatedFiles = "**/generated/**"
private val scriptsFiles = "**/*.kts"
```

Basically, the tool will start in the `projectSource` and recursively look for files that match `**/*.kt` or `**/*.kts`.

**Comment by AntsyLich** [comment]
Can you like make a convention plugin and apply it to per module? Don't really want to run the whole task with `:app`

**Comment by theolm** [comment]
@AntsyLich  Sure, I can add it pre-module. 
I just don't see the reason for using a `convention plugin`. do you have any strong feelings about it? I would rather modify my Kotlin script and apply it in the `subprojects` (like it was with the ktlint).

**Comment by AntsyLich** [comment]
I plan to migrate everything to convention plugin

**Comment by theolm** [comment]
@AntsyLich Let me know what you think.

File: buildSrc/build.gradle.kts
Comment (AntsyLich) [inline buildSrc/build.gradle.kts:6]: is this needed?

File: .github/workflows/build_pull_request.yml
Comment (AntsyLich) [inline .github/workflows/build_pull_request.yml:40]: can you add a newline?

File: gradle/compose.versions.toml
Comment (AntsyLich) [inline gradle/compose.versions.toml:27]: add newline

**Review by AntsyLich — CHANGES_REQUESTED** [review]

File: buildSrc/build.gradle.kts
Comment (theolm) [reply buildSrc/build.gradle.kts:6]: For now, Is required to access version catalog inside convention plugin. 
It seems that better support for version catalog will be release in future versions of Gradle.

Comment (stevenyomi) [reply buildSrc/build.gradle.kts:6]: Context: https://github.com/gradle/gradle/issues/15383#issuecomment-779893192

According to a reply in this issue, Gradle 8.5 has some [new support](https://docs.gradle.org/8.5/release-notes.html#catalog-precompiled) for this so you could probably consider using that.

File: gradle/compose.versions.toml
Comment (theolm) [reply gradle/compose.versions.toml:27]: Done

File: .github/workflows/build_pull_request.yml
Comment (theolm) [reply .github/workflows/build_pull_request.yml]: Done

File: buildSrc/build.gradle.kts
Comment (theolm) [reply buildSrc/build.gradle.kts:6]: @AntsyLich @stevenyomi Yeah I tried that, It looks worse to me. 
You can't access the variable directly, you need to find by string, which is prone to error. 

```kt
dependencies {
    versionCatalogs
        .named("libs")
        .findLibrary("detekt-rules-formatting")
        .ifPresent {
            detektPlugins(it)
        }

    versionCatalogs
        .named("libs")
        .findLibrary("detekt-rules-compose")
        .ifPresent {
            detektPlugins(it)
        }
}
```

It's up to you guys. I think using the import, for now, its a better solution.

Comment (theolm) [reply buildSrc/build.gradle.kts:6]: Should I make the change? What's the decision here?

Comment (stevenyomi) [reply buildSrc/build.gradle.kts:6]: I think this is acceptable:

```kotlin
dependencies {
    val libs = versionCatalogs.named("libs")
    detektPlugins(libs.findLibrary("detekt.rules.formatting").get())
    detektPlugins(libs.findLibrary("detekt.rules.compose").get())
}
```

I'm just providing context, and won't review or decide anything.

Comment (theolm) [reply buildSrc/build.gradle.kts:6]: @stevenyomi Good catch, I missed the `.get()` method. Your solution works and is better than using `ifPresent`... I still prefer importing the `libs class` and accessing the declarations directly though, it is safer and we don't rely on strings. Anyway, I'm good with both solutions.

@AntsyLich Let me know what you think.

Comment (AntsyLich) [reply buildSrc/build.gradle.kts:6]: Keep it as it is. I'll do some refactoring before merging

Comment (theolm) [reply buildSrc/build.gradle.kts:6]: What kind of refactoring? I'm open to requests.
I'm in a team transition period at work, so I have free time lol.

Btw, my idea here is to open another PR with Detekt auto-fixes after merging this one.

Comment (AntsyLich) [reply buildSrc/build.gradle.kts:6]: Gonna make something like https://github.com/android/nowinandroid/tree/main/build-logic. I've WIP and will merge this after that.

Comment (stevenyomi) [reply buildSrc/build.gradle.kts:6]: FYI the `build-logic` you linked is the same as the `buildSrc` directory. Now in Android makes use of binary plugins, and the thing we're talking about is for precompiled script plugins. Reference: https://proandroiddev.com/sharing-build-logic-with-kotlin-dsl-203274f73013#ac6a

Comment (AntsyLich) [reply buildSrc/build.gradle.kts:6]: I'm aware. Gonna move the scripts inside `build-logic`

Comment (stevenyomi) [reply buildSrc/build.gradle.kts:6]: I haven't dug very deep but it seems Gradle 8.x already had optimizations on `buildSrc`, so you'd probably need to research further what are the current drawbacks of `buildSrc`.

https://github.com/gradle/gradle/pull/22540
https://docs.gradle.org/8.0/release-notes.html#improvements-for-buildsrc-builds
https://github.com/android/nowinandroid/discussions/962

Comment (theolm) [reply buildSrc/build.gradle.kts:6]: @AntsyLich Good idea. I think it will be a good improvment.
I usually structure my projects like this:
https://github.com/theolm/kmm-template/tree/main/build-logic

Do you have something ready yet? otherwise I can take this refactor.

**Comment by theolm** [comment]
@AntsyLich Please ping me before merging this PR.
Several PRs have been merged since I opened this one, so I would like to update the baseline before the merge. This will avoid problems.

**Comment by AntsyLich** [comment]
I may have forgotten to ask you to rebase. Do that and ping me I'll merge it afterwards

**Comment by theolm** [comment]
@AntsyLich Done.

**Comment by AntsyLich** [comment]
> Optionally, it is possible to slim down the configuration file to only the changes from the default configuration, by applying the `buildUponDefaultConfig` option

We don't need that huge ass yml file do we? Also add https://github.com/mrmans0n/compose-rules (https://mrmans0n.github.io/compose-rules/detekt)

**Comment by theolm** [comment]
@AntsyLich
Ok, I'll replace the Compose rule set.
about the config file, are you sure you to slim down the file? I found very useful to have all the rules listed.

**Comment by AntsyLich** [comment]
it's fine to slim down the config file. chances are we will just keep following the default

**Comment by theolm** [comment]
@AntsyLich Done.

**Comment by AntsyLich** [comment]
According to https://mrmans0n.github.io/compose-rules/detekt/ you need to enable some rules 🤔 one of them also seem to supersede `FunctionNaming` if I'm not wrong

**Comment by theolm** [comment]
@AntsyLich The plugin rules are enabled by default after applying the plugin. I trimmed down the config files removing all the rules with default values (as you asked).

Here is the generated config file after applying the plugin.
Screenshot 2024-01-28 at 6 24 16 PM

You can also see that the baseline has several compose-rules suppressed.

**Review by AntsyLich — APPROVED** [review]

**Comment by AntsyLich** [comment]
Any reason why you didn't use detekt v1.23.4?

**Pull request merged**
